{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Accessing 10x RNA-seq gene expression data\n",
    "\n",
    "This notebook provides examples and functions for accessing the 10X expression matrix data stored in the ABC Atlas. These files require a large amount of memory to be available to load and analyize them if care is not taken. In this notebook, we present an example of how to access specific gene expressions from the data. The functions used below could be simplily parallized when processing data at scale, however, we leave them simple here.\n",
    "\n",
    "Care should still be taken not to attempt too load to many genes from the expression matrices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import List\n",
    "import os\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import anndata\n",
    "import time\n",
    "\n",
    "from abc_atlas_access.abc_atlas_cache.abc_project_cache import AbcProjectCache\n",
    "from abc_atlas_access.abc_atlas_cache.anndata_utils import get_gene_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will interact with the data using the **AbcProjectCache**. This cache object tracks which data has been downloaded and serves the path to the requsted data on disk. For metadata, the cache can also directly serve up a Pandas Dataframe. See the ``getting_started`` notebook for more details on using the cache including installing it if it has not already been.\n",
    "\n",
    "**Change the download_base variable to where you have downloaded the data in your system.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'releases/20240330/manifest.json'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "download_base = Path('/allen/programs/celltypes/workgroups/mousecelltypes/Rohan/dat/raw/abc_atlas')\n",
    "abc_cache = AbcProjectCache.from_s3_cache(download_base)\n",
    "abc_cache.current_manifest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene expression matrices\n",
    "\n",
    "The Whole Mouse Brain (WMB) and Whole Human Brain (WHB) datasets are formatted similarly. Each package is formatted as annadata h5ad files with minimal metadata. For each dataset, there are two h5ad files: one storing the raw counts and the other log2 normalization of the counts.\n",
    "\n",
    "To load the data by gene for either mouse or human dataset, we need to load two pieces of metadata, the ``cell``s table and the ``gene``s table in addition to our instantiated AbcProjectCache object. These metadata can be found in the directories WMB-10X and WHB-10Xv3 for mouse and human respectively. Below we use the human brain data in our example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47\n"
     ]
    }
   ],
   "source": [
    "ligands = [['Npy'],[ 'Sst'], ['Vip'], ['Tac2'], ['Cck'], ['Penk'], ['Crh'], ['Cort'], ['Tac1'], ['Pdyn'], ['Pthlh'], ['Pnoc'], ['Trh'], ['Grp'], ['Rln1'], ['Adcyap1'], ['Nts'], ['Nmb']]\n",
    "receptors = [[\"Npy1r\", \"Npy2r\", \"Npy5\"], [\"Sstr1\", \"Sstr2\", \"Sstr3\", \"Sstr4\" ], [\"Vipr1\", \"Vipr2\"], [\"Tacr3\"], [\"Cckbr\"], [\"Oprd1\", \"Oprm1\"], [\"Crhr1\", \"Crhr2\"], [\"Sstr1\", \"Sstr2\", \"Sstr3\", \"Sstr4\"], [\"Tacr1\"], [\"Oprd1\", \"Oprk1\", \"Oprm1\"], [\"Pth1r\"], [\"Oprl1\"], [\"Trhr\", \"Trhr2\"\n",
    "], [\"Grpr\"], [\"Rxfp1\", \"Rxfp2\", \"Rxfp3\"], [\"Adcyap1r1\", \"Vipr1\", \"Vipr2\"], [\"Ntsr1\", \"Ntsr2\"], [\"Nmbr\"]]\n",
    "\n",
    "lr_genes = ligands + receptors\n",
    "lr_genes = [item for sublist in lr_genes for item in sublist]\n",
    "lr_genes = list(set(lr_genes))\n",
    "print(len(lr_genes))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>mapped_ncbi_identifier</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000051951</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>X-linked Kx blood group related 4</td>\n",
       "      <td>NCBIGene:497097</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000089699</th>\n",
       "      <td>Gm1992</td>\n",
       "      <td>predicted gene 1992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102331</th>\n",
       "      <td>Gm19938</td>\n",
       "      <td>predicted gene, 19938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102343</th>\n",
       "      <td>Gm37381</td>\n",
       "      <td>predicted gene, 37381</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025900</th>\n",
       "      <td>Rp1</td>\n",
       "      <td>retinitis pigmentosa 1 (human)</td>\n",
       "      <td>NCBIGene:19888</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_symbol                               name  \\\n",
       "gene_identifier                                                     \n",
       "ENSMUSG00000051951        Xkr4  X-linked Kx blood group related 4   \n",
       "ENSMUSG00000089699      Gm1992                predicted gene 1992   \n",
       "ENSMUSG00000102331     Gm19938              predicted gene, 19938   \n",
       "ENSMUSG00000102343     Gm37381              predicted gene, 37381   \n",
       "ENSMUSG00000025900         Rp1     retinitis pigmentosa 1 (human)   \n",
       "\n",
       "                   mapped_ncbi_identifier comment  \n",
       "gene_identifier                                    \n",
       "ENSMUSG00000051951        NCBIGene:497097     NaN  \n",
       "ENSMUSG00000089699                    NaN     NaN  \n",
       "ENSMUSG00000102331                    NaN     NaN  \n",
       "ENSMUSG00000102343                    NaN     NaN  \n",
       "ENSMUSG00000025900         NCBIGene:19888     NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "MERFISH_genes = abc_cache.get_metadata_dataframe(directory='MERFISH-C57BL6J-638850', file_name='gene')\n",
    "MERFISH_genes = MERFISH_genes[~MERFISH_genes['gene_symbol'].str.startswith('Blank')]\n",
    "#display(MERFISH_genes.head())\n",
    "\n",
    "sc_genes = abc_cache.get_metadata_dataframe(directory='WMB-10X', file_name='gene').set_index('gene_identifier')\n",
    "display(sc_genes.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(528, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>mapped_ncbi_identifier</th>\n",
       "      <th>comment</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025905</th>\n",
       "      <td>Oprk1</td>\n",
       "      <td>opioid receptor, kappa 1</td>\n",
       "      <td>NCBIGene:18387</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033774</th>\n",
       "      <td>Npbwr1</td>\n",
       "      <td>neuropeptides B/W receptor 1</td>\n",
       "      <td>NCBIGene:226304</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033740</th>\n",
       "      <td>St18</td>\n",
       "      <td>suppression of tumorigenicity 18</td>\n",
       "      <td>NCBIGene:240690</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000042501</th>\n",
       "      <td>Cpa6</td>\n",
       "      <td>carboxypeptidase A6</td>\n",
       "      <td>NCBIGene:329093</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000042596</th>\n",
       "      <td>Tfap2d</td>\n",
       "      <td>transcription factor AP-2, delta</td>\n",
       "      <td>NCBIGene:226896</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_symbol                              name  \\\n",
       "gene_identifier                                                    \n",
       "ENSMUSG00000025905       Oprk1          opioid receptor, kappa 1   \n",
       "ENSMUSG00000033774      Npbwr1      neuropeptides B/W receptor 1   \n",
       "ENSMUSG00000033740        St18  suppression of tumorigenicity 18   \n",
       "ENSMUSG00000042501        Cpa6               carboxypeptidase A6   \n",
       "ENSMUSG00000042596      Tfap2d  transcription factor AP-2, delta   \n",
       "\n",
       "                   mapped_ncbi_identifier comment  \n",
       "gene_identifier                                    \n",
       "ENSMUSG00000025905         NCBIGene:18387     NaN  \n",
       "ENSMUSG00000033774        NCBIGene:226304     NaN  \n",
       "ENSMUSG00000033740        NCBIGene:240690     NaN  \n",
       "ENSMUSG00000042501        NCBIGene:329093     NaN  \n",
       "ENSMUSG00000042596        NCBIGene:226896     NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Keep only genes in SC_genes that are in MERFISH_genes[\"gene_symbol\"] or in lr_genes. Use the gene_symbol as the key\n",
    "MERFISH_genes_list = MERFISH_genes['gene_symbol'].tolist()\n",
    "mlr_genes = list(set(MERFISH_genes_list + lr_genes))\n",
    "\n",
    "target_genes = sc_genes[sc_genes['gene_symbol'].isin(mlr_genes)]\n",
    "print(target_genes.shape)\n",
    "display(target_genes.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(155383, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>barcoded_cell_sample_label</th>\n",
       "      <th>library_label</th>\n",
       "      <th>feature_matrix_label</th>\n",
       "      <th>entity</th>\n",
       "      <th>brain_section_label</th>\n",
       "      <th>library_method</th>\n",
       "      <th>region_of_interest_acronym</th>\n",
       "      <th>donor_label</th>\n",
       "      <th>donor_genotype</th>\n",
       "      <th>donor_sex</th>\n",
       "      <th>dataset_label</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>cluster_alias</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CATGGTATCGCATTAG-477_A01</th>\n",
       "      <td>CATGGTATCGCATTAG</td>\n",
       "      <td>477_A01</td>\n",
       "      <td>L8TX_210107_01_C08</td>\n",
       "      <td>WMB-10Xv3-Isocortex-1</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv3</td>\n",
       "      <td>VIS</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-557707</td>\n",
       "      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>\n",
       "      <td>F</td>\n",
       "      <td>WMB-10Xv3</td>\n",
       "      <td>11.900758</td>\n",
       "      <td>-3.864682</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACAGGCGTTCTGACA-480_A04</th>\n",
       "      <td>CACAGGCGTTCTGACA</td>\n",
       "      <td>480_A04</td>\n",
       "      <td>L8TX_210107_01_F09</td>\n",
       "      <td>WMB-10Xv3-Isocortex-1</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv3</td>\n",
       "      <td>VIS</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-556846</td>\n",
       "      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>WMB-10Xv3</td>\n",
       "      <td>11.421314</td>\n",
       "      <td>-3.986752</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGGAAGCAGCATCTA-480_A04</th>\n",
       "      <td>GTGGAAGCAGCATCTA</td>\n",
       "      <td>480_A04</td>\n",
       "      <td>L8TX_210107_01_F09</td>\n",
       "      <td>WMB-10Xv3-Isocortex-1</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv3</td>\n",
       "      <td>VIS</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-556846</td>\n",
       "      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>WMB-10Xv3</td>\n",
       "      <td>11.917944</td>\n",
       "      <td>-3.799100</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              cell_barcode barcoded_cell_sample_label  \\\n",
       "cell_label                                                              \n",
       "CATGGTATCGCATTAG-477_A01  CATGGTATCGCATTAG                    477_A01   \n",
       "CACAGGCGTTCTGACA-480_A04  CACAGGCGTTCTGACA                    480_A04   \n",
       "GTGGAAGCAGCATCTA-480_A04  GTGGAAGCAGCATCTA                    480_A04   \n",
       "\n",
       "                               library_label   feature_matrix_label entity  \\\n",
       "cell_label                                                                   \n",
       "CATGGTATCGCATTAG-477_A01  L8TX_210107_01_C08  WMB-10Xv3-Isocortex-1   cell   \n",
       "CACAGGCGTTCTGACA-480_A04  L8TX_210107_01_F09  WMB-10Xv3-Isocortex-1   cell   \n",
       "GTGGAAGCAGCATCTA-480_A04  L8TX_210107_01_F09  WMB-10Xv3-Isocortex-1   cell   \n",
       "\n",
       "                          brain_section_label library_method  \\\n",
       "cell_label                                                     \n",
       "CATGGTATCGCATTAG-477_A01                  NaN          10Xv3   \n",
       "CACAGGCGTTCTGACA-480_A04                  NaN          10Xv3   \n",
       "GTGGAAGCAGCATCTA-480_A04                  NaN          10Xv3   \n",
       "\n",
       "                         region_of_interest_acronym  \\\n",
       "cell_label                                            \n",
       "CATGGTATCGCATTAG-477_A01                        VIS   \n",
       "CACAGGCGTTCTGACA-480_A04                        VIS   \n",
       "GTGGAAGCAGCATCTA-480_A04                        VIS   \n",
       "\n",
       "                                           donor_label  \\\n",
       "cell_label                                               \n",
       "CATGGTATCGCATTAG-477_A01  Snap25-IRES2-Cre;Ai14-557707   \n",
       "CACAGGCGTTCTGACA-480_A04  Snap25-IRES2-Cre;Ai14-556846   \n",
       "GTGGAAGCAGCATCTA-480_A04  Snap25-IRES2-Cre;Ai14-556846   \n",
       "\n",
       "                                                donor_genotype donor_sex  \\\n",
       "cell_label                                                                 \n",
       "CATGGTATCGCATTAG-477_A01  Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt         F   \n",
       "CACAGGCGTTCTGACA-480_A04  Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt         M   \n",
       "GTGGAAGCAGCATCTA-480_A04  Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt         M   \n",
       "\n",
       "                         dataset_label          x         y  cluster_alias  \n",
       "cell_label                                                                  \n",
       "CATGGTATCGCATTAG-477_A01     WMB-10Xv3  11.900758 -3.864682            102  \n",
       "CACAGGCGTTCTGACA-480_A04     WMB-10Xv3  11.421314 -3.986752            102  \n",
       "GTGGAAGCAGCATCTA-480_A04     WMB-10Xv3  11.917944 -3.799100            102  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# metadata for cells for which (library_method is 10Xv3) AND (region_of_interest_acronym is either VIS or MOp)\n",
    "sc_cells = abc_cache.get_metadata_dataframe(directory='WMB-10X', file_name='cell_metadata').set_index('cell_label')\n",
    "\n",
    "sc_cells = sc_cells[(sc_cells[\"library_method\"] == \"10Xv3\") & (sc_cells[\"region_of_interest_acronym\"].isin([\"VIS\", \"MOp\"]))]\n",
    "sc_cells[[\"region_of_interest_acronym\",\"library_method\"]].value_counts()\n",
    "\n",
    "print(sc_cells.shape)\n",
    "display(sc_cells.head(3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compiling MOp and VIS scRNAseq data, including the annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as ad\n",
    "import toml\n",
    "\n",
    "paths = toml.load('/allen/programs/celltypes/workgroups/mousecelltypes/Rohan/code/abcex/config.toml')\n",
    "expr_data = ad.read_h5ad(paths['data']+'expression_matrices/WMB-10Xv3/20230630/WMB-10Xv3-Isocortex-1-log2.h5ad', backed='r')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "region_of_interest_acronym  library_method\n",
       "MOp                         10Xv3             124501\n",
       "VIS                         10Xv3              30882\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# all metadata\n",
    "sc_cells = abc_cache.get_metadata_dataframe(directory='WMB-10X', file_name='cell_metadata').set_index('cell_label')\n",
    "sc_cells = sc_cells[(sc_cells[\"library_method\"] == \"10Xv3\") & (sc_cells[\"region_of_interest_acronym\"].isin([\"VIS\", \"MOp\"]))]\n",
    "sc_cells[[\"region_of_interest_acronym\",\"library_method\"]].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "expression data:  228229\n",
      "filtered metadata:  155383\n",
      "filtered metadata cells that have expression data:  90714\n"
     ]
    }
   ],
   "source": [
    "overlapping_cells = np.intersect1d(expr_data.obs.index.values, sc_cells.index.values)\n",
    "print(\"expression data: \", np.size(expr_data.obs.index.values))\n",
    "print(\"filtered metadata: \", np.size(sc_cells.index.values))\n",
    "print(\"filtered metadata cells that have expression data: \", np.size(overlapping_cells))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(90714, 528)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>library_label</th>\n",
       "      <th>anatomical_division_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAACAAGG-480_A04</th>\n",
       "      <td>AAACCCAAGAACAAGG</td>\n",
       "      <td>L8TX_210107_01_F09</td>\n",
       "      <td>Isocortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAACTTCC-198.1_A01</th>\n",
       "      <td>AAACCCAAGAACTTCC</td>\n",
       "      <td>L8TX_200106_01_G07</td>\n",
       "      <td>Isocortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAGTATC-116_C01</th>\n",
       "      <td>AAACCCAAGAAGTATC</td>\n",
       "      <td>L8TX_190430_01_C08</td>\n",
       "      <td>Isocortex</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                cell_barcode       library_label  \\\n",
       "cell_label                                                         \n",
       "AAACCCAAGAACAAGG-480_A04    AAACCCAAGAACAAGG  L8TX_210107_01_F09   \n",
       "AAACCCAAGAACTTCC-198.1_A01  AAACCCAAGAACTTCC  L8TX_200106_01_G07   \n",
       "AAACCCAAGAAGTATC-116_C01    AAACCCAAGAAGTATC  L8TX_190430_01_C08   \n",
       "\n",
       "                           anatomical_division_label  \n",
       "cell_label                                            \n",
       "AAACCCAAGAACAAGG-480_A04                   Isocortex  \n",
       "AAACCCAAGAACTTCC-198.1_A01                 Isocortex  \n",
       "AAACCCAAGAAGTATC-116_C01                   Isocortex  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# get expr.obs where the index values match sc_cells.index\n",
    "# adata = expr_data.to_memory()\n",
    "\n",
    "adata = expr_data[expr_data.obs.loc[overlapping_cells].index, target_genes.index].to_memory()\n",
    "print(adata.shape)\n",
    "\n",
    "# annotations for type/region etc. are missing from this file. \n",
    "display(adata.obs.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neurotransmitter</th>\n",
       "      <th>class</th>\n",
       "      <th>subclass</th>\n",
       "      <th>supertype</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_alias</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Glut</td>\n",
       "      <td>01 IT-ET Glut</td>\n",
       "      <td>018 L2 IT PPP-APr Glut</td>\n",
       "      <td>0082 L2 IT PPP-APr Glut_3</td>\n",
       "      <td>0326 L2 IT PPP-APr Glut_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Glut</td>\n",
       "      <td>01 IT-ET Glut</td>\n",
       "      <td>018 L2 IT PPP-APr Glut</td>\n",
       "      <td>0082 L2 IT PPP-APr Glut_3</td>\n",
       "      <td>0327 L2 IT PPP-APr Glut_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Glut</td>\n",
       "      <td>01 IT-ET Glut</td>\n",
       "      <td>018 L2 IT PPP-APr Glut</td>\n",
       "      <td>0081 L2 IT PPP-APr Glut_2</td>\n",
       "      <td>0322 L2 IT PPP-APr Glut_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              neurotransmitter          class                subclass  \\\n",
       "cluster_alias                                                           \n",
       "1                         Glut  01 IT-ET Glut  018 L2 IT PPP-APr Glut   \n",
       "2                         Glut  01 IT-ET Glut  018 L2 IT PPP-APr Glut   \n",
       "3                         Glut  01 IT-ET Glut  018 L2 IT PPP-APr Glut   \n",
       "\n",
       "                               supertype                    cluster  \n",
       "cluster_alias                                                        \n",
       "1              0082 L2 IT PPP-APr Glut_3  0326 L2 IT PPP-APr Glut_3  \n",
       "2              0082 L2 IT PPP-APr Glut_3  0327 L2 IT PPP-APr Glut_3  \n",
       "3              0081 L2 IT PPP-APr Glut_2  0322 L2 IT PPP-APr Glut_2  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neurotransmitter_color</th>\n",
       "      <th>class_color</th>\n",
       "      <th>subclass_color</th>\n",
       "      <th>supertype_color</th>\n",
       "      <th>cluster_color</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_alias</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#FA0087</td>\n",
       "      <td>#0F6632</td>\n",
       "      <td>#266DFF</td>\n",
       "      <td>#64661F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#FA0087</td>\n",
       "      <td>#0F6632</td>\n",
       "      <td>#266DFF</td>\n",
       "      <td>#CCA73D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#FA0087</td>\n",
       "      <td>#0F6632</td>\n",
       "      <td>#002BCC</td>\n",
       "      <td>#99000D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              neurotransmitter_color class_color subclass_color  \\\n",
       "cluster_alias                                                     \n",
       "1                            #2B93DF     #FA0087        #0F6632   \n",
       "2                            #2B93DF     #FA0087        #0F6632   \n",
       "3                            #2B93DF     #FA0087        #0F6632   \n",
       "\n",
       "              supertype_color cluster_color  \n",
       "cluster_alias                                \n",
       "1                     #266DFF       #64661F  \n",
       "2                     #266DFF       #CCA73D  \n",
       "3                     #002BCC       #99000D  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label</th>\n",
       "      <th>name</th>\n",
       "      <th>region_of_interest_order</th>\n",
       "      <th>region_of_interest_color</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acronym</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MO-FRP</th>\n",
       "      <td>WMB-MO-FRP</td>\n",
       "      <td>Somatomotor - Frontal pole</td>\n",
       "      <td>0</td>\n",
       "      <td>#3DCC7C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MOp</th>\n",
       "      <td>WMB-MOp</td>\n",
       "      <td>Primary motor area</td>\n",
       "      <td>1</td>\n",
       "      <td>#179968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SS-GU-VISC</th>\n",
       "      <td>WMB-SS-GU-VISC</td>\n",
       "      <td>Somatosensory/gustatory/visceral areas</td>\n",
       "      <td>2</td>\n",
       "      <td>#2E8599</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     label                                    name  \\\n",
       "acronym                                                              \n",
       "MO-FRP          WMB-MO-FRP              Somatomotor - Frontal pole   \n",
       "MOp                WMB-MOp                      Primary motor area   \n",
       "SS-GU-VISC  WMB-SS-GU-VISC  Somatosensory/gustatory/visceral areas   \n",
       "\n",
       "            region_of_interest_order region_of_interest_color  \n",
       "acronym                                                        \n",
       "MO-FRP                             0                  #3DCC7C  \n",
       "MOp                                1                  #179968  \n",
       "SS-GU-VISC                         2                  #2E8599  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# get the cluster annotations\n",
    "cluster_details = abc_cache.get_metadata_dataframe(\n",
    "    directory='WMB-taxonomy',\n",
    "    file_name='cluster_to_cluster_annotation_membership_pivoted',\n",
    "    keep_default_na=False\n",
    ")\n",
    "cluster_details.set_index('cluster_alias', inplace=True)\n",
    "display(cluster_details.head(3))\n",
    "\n",
    "# get the cluster colors\n",
    "cluster_colors = abc_cache.get_metadata_dataframe(\n",
    "    directory='WMB-taxonomy',\n",
    "    file_name='cluster_to_cluster_annotation_membership_color'\n",
    ")\n",
    "cluster_colors.set_index('cluster_alias', inplace=True)\n",
    "display(cluster_colors.head(3))\n",
    "\n",
    "# get the roi annotations\n",
    "roi = abc_cache.get_metadata_dataframe(directory='WMB-10X', file_name='region_of_interest_metadata')\n",
    "roi.set_index('acronym', inplace=True)\n",
    "roi.rename(columns={'order': 'region_of_interest_order',\n",
    "                    'color_hex_triplet': 'region_of_interest_color'},\n",
    "           inplace=True)\n",
    "display(roi.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>barcoded_cell_sample_label</th>\n",
       "      <th>library_label</th>\n",
       "      <th>feature_matrix_label</th>\n",
       "      <th>entity</th>\n",
       "      <th>brain_section_label</th>\n",
       "      <th>library_method</th>\n",
       "      <th>region_of_interest_acronym</th>\n",
       "      <th>donor_label</th>\n",
       "      <th>donor_genotype</th>\n",
       "      <th>...</th>\n",
       "      <th>subclass</th>\n",
       "      <th>supertype</th>\n",
       "      <th>cluster</th>\n",
       "      <th>neurotransmitter_color</th>\n",
       "      <th>class_color</th>\n",
       "      <th>subclass_color</th>\n",
       "      <th>supertype_color</th>\n",
       "      <th>cluster_color</th>\n",
       "      <th>region_of_interest_order</th>\n",
       "      <th>region_of_interest_color</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAACAAGG-480_A04</th>\n",
       "      <td>AAACCCAAGAACAAGG</td>\n",
       "      <td>480_A04</td>\n",
       "      <td>L8TX_210107_01_F09</td>\n",
       "      <td>WMB-10Xv3-Isocortex-1</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv3</td>\n",
       "      <td>VIS</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-556846</td>\n",
       "      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>\n",
       "      <td>...</td>\n",
       "      <td>007 L2/3 IT CTX Glut</td>\n",
       "      <td>0030 L2/3 IT CTX Glut_2</td>\n",
       "      <td>0109 L2/3 IT CTX Glut_2</td>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#FA0087</td>\n",
       "      <td>#E9530F</td>\n",
       "      <td>#CC7A8F</td>\n",
       "      <td>#AA7ACC</td>\n",
       "      <td>7</td>\n",
       "      <td>#0059CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAACTTCC-198.1_A01</th>\n",
       "      <td>AAACCCAAGAACTTCC</td>\n",
       "      <td>198.1_A01</td>\n",
       "      <td>L8TX_200106_01_G07</td>\n",
       "      <td>WMB-10Xv3-Isocortex-1</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv3</td>\n",
       "      <td>MOp</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-500199</td>\n",
       "      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>\n",
       "      <td>...</td>\n",
       "      <td>006 L4/5 IT CTX Glut</td>\n",
       "      <td>0023 L4/5 IT CTX Glut_1</td>\n",
       "      <td>0069 L4/5 IT CTX Glut_1</td>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#FA0087</td>\n",
       "      <td>#3DCCB7</td>\n",
       "      <td>#3B9900</td>\n",
       "      <td>#997145</td>\n",
       "      <td>1</td>\n",
       "      <td>#179968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGAAGTATC-116_C01</th>\n",
       "      <td>AAACCCAAGAAGTATC</td>\n",
       "      <td>116_C01</td>\n",
       "      <td>L8TX_190430_01_C08</td>\n",
       "      <td>WMB-10Xv3-Isocortex-1</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv3</td>\n",
       "      <td>MOp</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-457911</td>\n",
       "      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>\n",
       "      <td>...</td>\n",
       "      <td>007 L2/3 IT CTX Glut</td>\n",
       "      <td>0030 L2/3 IT CTX Glut_2</td>\n",
       "      <td>0109 L2/3 IT CTX Glut_2</td>\n",
       "      <td>#2B93DF</td>\n",
       "      <td>#FA0087</td>\n",
       "      <td>#E9530F</td>\n",
       "      <td>#CC7A8F</td>\n",
       "      <td>#AA7ACC</td>\n",
       "      <td>1</td>\n",
       "      <td>#179968</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                cell_barcode barcoded_cell_sample_label  \\\n",
       "cell_label                                                                \n",
       "AAACCCAAGAACAAGG-480_A04    AAACCCAAGAACAAGG                    480_A04   \n",
       "AAACCCAAGAACTTCC-198.1_A01  AAACCCAAGAACTTCC                  198.1_A01   \n",
       "AAACCCAAGAAGTATC-116_C01    AAACCCAAGAAGTATC                    116_C01   \n",
       "\n",
       "                                 library_label   feature_matrix_label entity  \\\n",
       "cell_label                                                                     \n",
       "AAACCCAAGAACAAGG-480_A04    L8TX_210107_01_F09  WMB-10Xv3-Isocortex-1   cell   \n",
       "AAACCCAAGAACTTCC-198.1_A01  L8TX_200106_01_G07  WMB-10Xv3-Isocortex-1   cell   \n",
       "AAACCCAAGAAGTATC-116_C01    L8TX_190430_01_C08  WMB-10Xv3-Isocortex-1   cell   \n",
       "\n",
       "                            brain_section_label library_method  \\\n",
       "cell_label                                                       \n",
       "AAACCCAAGAACAAGG-480_A04                    NaN          10Xv3   \n",
       "AAACCCAAGAACTTCC-198.1_A01                  NaN          10Xv3   \n",
       "AAACCCAAGAAGTATC-116_C01                    NaN          10Xv3   \n",
       "\n",
       "                           region_of_interest_acronym  \\\n",
       "cell_label                                              \n",
       "AAACCCAAGAACAAGG-480_A04                          VIS   \n",
       "AAACCCAAGAACTTCC-198.1_A01                        MOp   \n",
       "AAACCCAAGAAGTATC-116_C01                          MOp   \n",
       "\n",
       "                                             donor_label  \\\n",
       "cell_label                                                 \n",
       "AAACCCAAGAACAAGG-480_A04    Snap25-IRES2-Cre;Ai14-556846   \n",
       "AAACCCAAGAACTTCC-198.1_A01  Snap25-IRES2-Cre;Ai14-500199   \n",
       "AAACCCAAGAAGTATC-116_C01    Snap25-IRES2-Cre;Ai14-457911   \n",
       "\n",
       "                                                  donor_genotype  ...  \\\n",
       "cell_label                                                        ...   \n",
       "AAACCCAAGAACAAGG-480_A04    Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt  ...   \n",
       "AAACCCAAGAACTTCC-198.1_A01  Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt  ...   \n",
       "AAACCCAAGAAGTATC-116_C01    Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt  ...   \n",
       "\n",
       "                                        subclass                supertype  \\\n",
       "cell_label                                                                  \n",
       "AAACCCAAGAACAAGG-480_A04    007 L2/3 IT CTX Glut  0030 L2/3 IT CTX Glut_2   \n",
       "AAACCCAAGAACTTCC-198.1_A01  006 L4/5 IT CTX Glut  0023 L4/5 IT CTX Glut_1   \n",
       "AAACCCAAGAAGTATC-116_C01    007 L2/3 IT CTX Glut  0030 L2/3 IT CTX Glut_2   \n",
       "\n",
       "                                            cluster  neurotransmitter_color  \\\n",
       "cell_label                                                                    \n",
       "AAACCCAAGAACAAGG-480_A04    0109 L2/3 IT CTX Glut_2                 #2B93DF   \n",
       "AAACCCAAGAACTTCC-198.1_A01  0069 L4/5 IT CTX Glut_1                 #2B93DF   \n",
       "AAACCCAAGAAGTATC-116_C01    0109 L2/3 IT CTX Glut_2                 #2B93DF   \n",
       "\n",
       "                            class_color subclass_color supertype_color  \\\n",
       "cell_label                                                               \n",
       "AAACCCAAGAACAAGG-480_A04        #FA0087        #E9530F         #CC7A8F   \n",
       "AAACCCAAGAACTTCC-198.1_A01      #FA0087        #3DCCB7         #3B9900   \n",
       "AAACCCAAGAAGTATC-116_C01        #FA0087        #E9530F         #CC7A8F   \n",
       "\n",
       "                           cluster_color region_of_interest_order  \\\n",
       "cell_label                                                          \n",
       "AAACCCAAGAACAAGG-480_A04         #AA7ACC                        7   \n",
       "AAACCCAAGAACTTCC-198.1_A01       #997145                        1   \n",
       "AAACCCAAGAAGTATC-116_C01         #AA7ACC                        1   \n",
       "\n",
       "                           region_of_interest_color  \n",
       "cell_label                                           \n",
       "AAACCCAAGAACAAGG-480_A04                    #0059CC  \n",
       "AAACCCAAGAACTTCC-198.1_A01                  #179968  \n",
       "AAACCCAAGAAGTATC-116_C01                    #179968  \n",
       "\n",
       "[3 rows x 27 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "obs_new_ = sc_cells.loc[overlapping_cells]\n",
    "obs_new = obs_new_.join(cluster_details, on='cluster_alias')\n",
    "obs_new = obs_new.join(cluster_colors, on='cluster_alias')\n",
    "obs_new = obs_new.join(roi[['region_of_interest_order', 'region_of_interest_color']], on='region_of_interest_acronym')\n",
    "display(obs_new.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>subclass</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>030 L6 CT CTX Glut</th>\n",
       "      <td>23562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>006 L4/5 IT CTX Glut</th>\n",
       "      <td>23177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>007 L2/3 IT CTX Glut</th>\n",
       "      <td>19084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>032 L5 NP CTX Glut</th>\n",
       "      <td>5445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>004 L6 IT CTX Glut</th>\n",
       "      <td>4019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>053 Sst Gaba</th>\n",
       "      <td>3123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>046 Vip Gaba</th>\n",
       "      <td>2643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>005 L5 IT CTX Glut</th>\n",
       "      <td>2362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>049 Lamp5 Gaba</th>\n",
       "      <td>1907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>029 L6b CTX Glut</th>\n",
       "      <td>1220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>022 L5 ET CTX Glut</th>\n",
       "      <td>888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>052 Pvalb Gaba</th>\n",
       "      <td>793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>047 Sncg Gaba</th>\n",
       "      <td>752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327 Oligo NN</th>\n",
       "      <td>585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319 Astro-TE NN</th>\n",
       "      <td>278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>056 Sst Chodl Gaba</th>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333 Endo NN</th>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326 OPC NN</th>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334 Microglia NN</th>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>001 CLA-EPd-CTX Car3 Glut</th>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331 Peri NN</th>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>051 Pvalb chandelier Gaba</th>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>050 Lamp5 Lhx6 Gaba</th>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332 SMC NN</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335 BAM NN</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330 VLMC NN</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>036 HPF CR Glut</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>045 OB-STR-CTX Inh IMN</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>063 STR D1 Sema5a Gaba</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>002 IT EP-CLA Glut</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>041 OB-in Frmd7 Gaba</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>044 OB Dopa-Gaba</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337 DC NN</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329 ABC NN</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338 Lymphoid NN</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>025 CA2-FC-IG Glut</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>054 STR Prox1 Lhx6 Gaba</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336 Monocytes NN</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>039 OB Meis2 Thsd7b Gaba</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>003 L5/6 IT TPE-ENT Glut</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>061 STR D1 Gaba</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>042 OB-out Frmd7 Gaba</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           count\n",
       "subclass                        \n",
       "030 L6 CT CTX Glut         23562\n",
       "006 L4/5 IT CTX Glut       23177\n",
       "007 L2/3 IT CTX Glut       19084\n",
       "032 L5 NP CTX Glut          5445\n",
       "004 L6 IT CTX Glut          4019\n",
       "053 Sst Gaba                3123\n",
       "046 Vip Gaba                2643\n",
       "005 L5 IT CTX Glut          2362\n",
       "049 Lamp5 Gaba              1907\n",
       "029 L6b CTX Glut            1220\n",
       "022 L5 ET CTX Glut           888\n",
       "052 Pvalb Gaba               793\n",
       "047 Sncg Gaba                752\n",
       "327 Oligo NN                 585\n",
       "319 Astro-TE NN              278\n",
       "056 Sst Chodl Gaba           163\n",
       "333 Endo NN                  160\n",
       "326 OPC NN                   156\n",
       "334 Microglia NN             123\n",
       "001 CLA-EPd-CTX Car3 Glut     72\n",
       "331 Peri NN                   49\n",
       "051 Pvalb chandelier Gaba     39\n",
       "050 Lamp5 Lhx6 Gaba           30\n",
       "332 SMC NN                    15\n",
       "335 BAM NN                    15\n",
       "330 VLMC NN                   11\n",
       "036 HPF CR Glut                7\n",
       "045 OB-STR-CTX Inh IMN         6\n",
       "063 STR D1 Sema5a Gaba         6\n",
       "002 IT EP-CLA Glut             4\n",
       "041 OB-in Frmd7 Gaba           3\n",
       "044 OB Dopa-Gaba               3\n",
       "337 DC NN                      3\n",
       "329 ABC NN                     2\n",
       "338 Lymphoid NN                2\n",
       "025 CA2-FC-IG Glut             1\n",
       "054 STR Prox1 Lhx6 Gaba        1\n",
       "336 Monocytes NN               1\n",
       "039 OB Meis2 Thsd7b Gaba       1\n",
       "003 L5/6 IT TPE-ENT Glut       1\n",
       "061 STR D1 Gaba                1\n",
       "042 OB-out Frmd7 Gaba          1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# checks\n",
    "assert np.all(obs_new.index == adata.obs.index), \"metadata order is same as adata order\"\n",
    "\n",
    "# replace the metadata with additional annotations\n",
    "adata.obs = obs_new\n",
    "\n",
    "display(adata.obs[\"subclass\"].value_counts().to_frame())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad('/allen/programs/celltypes/workgroups/mousecelltypes/Donghyun/data/Imputation/10Xv3_VIS_MOp_log2.h5ad')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
